Add registry inspector to the HTTP server (#50207)
elastic/beats: docs/reference/filebeat/http-endpoint.md

`http.pprof.mutex_profile_rate`
:   (Optional) `mutex_profile_rate` controls the fraction of mutex contention events that are reported in the mutex profile available from `/debug/pprof/mutex`. On average 1/rate events are reported. To turn off profiling entirely, pass rate 0. The default value is 0.

> `http.state_inspector.enabled`

Reviewer: Both the config option name and URL path look like ordinary, supported features. They sit next to `http.enabled`, `http.host`, `http.pprof.enabled`, etc. Nothing in the name distinguishes "internal debugger with no auth" from "stable knob."

Could we rename to:

- Config key: `http.debug.state_inspector.enabled`
- URL prefix: `/debug/store-inspector/states[.html]`

This mirrors Go's own `/debug/pprof/` convention (which we already use), gives us a place for future debug-only endpoints to land.

If you'd rather make the warning even more explicit, `http.debug.unsafe_state_inspector.enabled` works too.

Reviewer 2: +1

Reviewer 3: +1

Author: I think `unsafe_state_inspector.enabled` is a step a bit too far, but I like the idea to put under `debug`.

Author: Addressed in https://github.com/elastic/beats/pull/50207/commits/8651fc5f84f85aa9bab0577b1ac7ba7c93ec8982

> :   (Optional) Enable the state store inspector. This is a debugging tool intended for development and troubleshooting only. **Do not enable in production:** the inspector allows reading and deleting state entries, which may cause data loss or duplicate processing. When enabled, the internal state store used by the beat is exposed via HTTP for inspection and manipulation. Default is `false`. See [State Inspector](#state-inspector) for details.

Reviewer 3: This doc is from Filebeat, so we can be specific:

```suggestion
:   (Optional) Enable the state store inspector. This is a debugging tool intended for development and troubleshooting only. **Do not enable in production:** the inspector allows reading and deleting state entries, which may cause data loss or duplicate processing. When enabled, the internal state store used by Filebeat is exposed via HTTP for inspection and manipulation. Default is `false`. See [State Inspector](#state-inspector) for details.
```

Author: The wording was replaced in https://github.com/elastic/beats/pull/50207/commits/6df6b3814d30738a277956364bd1d22ff4662f4c

`http.pprof.mutex_profile_rate`
:   (Optional) `mutex_profile_rate` controls the fraction of mutex contention events that are reported in the mutex profile available from `/debug/pprof/mutex`. On average 1/rate events are reported. To turn off profiling entirely, pass rate 0. The default value is 0.

`http.state_inspector.enabled`

> :   (Optional) Enable the state store inspector. When enabled, the internal state store used by the beat is exposed via HTTP for inspection and manipulation. Default is `false`. See [State Inspector](#state-inspector) for details.

Reviewer 4: ```suggestion
:   (Optional) Enable the state store inspector. When enabled, the internal state store used by the beat is exposed through HTTP for inspection and manipulation. Default is `false`. Refer to [State Inspector](#state-inspector) for details.
```

Reviewer 4: If this is a new 9.5 feature, please add applies_to tags.

Author: Added the tag in https://github.com/elastic/beats/pull/50207/commits/0d1bf217f2ae0b2ee5a078ca00b8ad261bc4aaba

A request may optionally specify a `type` query parameter to request metrics for a specific type of input. And `pretty` may be included to have the returned JSON be pretty formatted.

```sh
curl 'http://localhost:5066/inputs/'
curl 'http://localhost:5066/inputs/?pretty'
curl 'http://localhost:5066/inputs/?type=aws-s3&pretty'
```

## State Inspector [state-inspector]

```{applies_to}

> stack: preview 9.5

Reviewer 2: [NIT]
I think it could be "lower", I believe we should use the most unstable/no guarantees possible. But not a blocker

Author: @AndersonQ what do you mean by "lower"? The lifecycle value? The supported values are restricted to: `preview`, `beta`, `ga`, `deprecated`, `removed`.

See https://elastic.github.io/docs-builder/syntax/applies/

Reviewer 2: Oh, my bad, preview is it then :)

curl 'http://localhost:5066/inputs/?type=aws-s3&pretty'
```

## State Inspector [state-inspector]

`/store-inspector/states` returns a JSON array of all key-value pairs in Filebeat's internal state store. `/store-inspector/states.html` serves a web UI for browsing and deleting individual state entries.

These endpoints are only available when `http.state_inspector.enabled` is set to `true`.

```js
curl -XGET 'localhost:5066/store-inspector/states?pretty'
```

To delete a specific state entry:

```js
curl -XDELETE 'localhost:5066/store-inspector/states/<key>'

> ```

Reviewer 2: [NIT]

`js`? isn't it `shell`?

Author: 🤦‍♂️
Fixed in https://github.com/elastic/beats/pull/50207/commits/0d1bf217f2ae0b2ee5a078ca00b8ad261bc4aaba
